Added `Print Page As PDF` Keyword

emanlove opened the pull request:
Some initial ideas and sketches for the `Print Page As PDF` keyword. I wanted to seewhat it may look like and how it may work. So this is just some preliminary ideas. Seeing a need to handle Print Options, need to document options and what are valid values, how to deal with the return encoded base64 string, etc.

emanlove committed:
Initial commit and sketchings of `Print Page As PDF` Keyword

Some initial ideas and sketches for the `Print Page As PDF` keyword. I
wanted to seewhat it may look like and how it may work. So this is just
some preliminary ideas. Seeing a need to handle Print Options, need to
document options and what are valid values, how to deal with the return
encoded base64 string, etc.

emanlove commented on src/SeleniumLibrary/keywords/screenshot.py:
```diff
@@ -235,3 +237,20 @@ def _embed_to_log_as_file(self, path, width):
+
+        base64code  = self.driver.print_page(print_options)
+        pdfdata = b64decode(base64code)
+        with open('test.pdf', mode='wb') as pdf:
```
This is very preliminary. For example, here I just use a test file. I expect a fair more code and allowing the user to specify the optional filename or filename syntax when printing page as pdf.

yuriverweij replied:
That was the first thing I noticed, thanks for clarifying.

yuriverweij replied:
Would it make sense to be able to set a default set of PrintOptions? Meaning the user can set the PrintOptions once, and those settings will be used every time the print_page_as_pdf will be called?
Something like with setting the timeouts  
`${old_value}  Set Print to PDF options  scale=0.5 password=${pdfpassword}`
or something like that

emanlove replied:
It does makes sense to have a context for print options. For the initial release I am going to leave that out though.

emanlove commented on atest/acceptance/keywords/print_page.robot:
```diff
@@ -0,0 +1,18 @@
+Print Page As PDF Without Print Options
+    Print Page As PDF
+
+Provide Print Options From Module
```
With this test I just wanted to do a preliminary exploration of how to set print options. A while back for when one opens a browser there were ways to customize how the browser was started. To do this one had to do something similar, which is go into the python module and make an instance of the python class of options. Then add to that robot variable as class instance with a method on that class to add attributes. It was fairly ugly and the syntax was all python. But doable. Tatu really cleaned that up with just having a string of options. I wanted to see if one could add print options that old school way. Here I left in my failed attempts as a reminder of what failed and what worked so if we do explain it with this then I can say what and what not to do.

emanlove commented on src/SeleniumLibrary/keywords/screenshot.py:
```diff
@@ -235,3 +237,20 @@ def _embed_to_log_as_file(self, path, width):
         )
+    
+    @keyword
+    # def print_page_as_pdf(self, print_options: Optional[PrintOptions]=None,):
```
Trying to figure out the proper type hints and type conversion signature for this method ..

emanlove replied:
Initially accepting the all the print options within keyword arguments.

yuriverweij reviewed (commented):
Haven't been able to try this yet, but added some thoughts/comments

emanlove committed:
Switched from single print_options argument to individual options

yuriverweij commented on src/SeleniumLibrary/keywords/screenshot.py:
```diff
@@ -241,16 +241,64 @@ def _embed_to_log_as_file(self, path, width):
     # def print_page_as_pdf(self, print_options: Optional[PrintOptions]=None,):
     # def print_page_as_pdf(self, print_options: Optional[Union[PrintOptions, dict]]=None):
-    def print_page_as_pdf(self, print_options: Union[PrintOptions, dict, None]=None):
+    # def print_page_as_pdf(self, print_options: Union[PrintOptions, dict, None]=None):
```
comments can be removed

emanlove replied:
Removed. Added some more, and eventually removed those as well. [Laughing]

emanlove committed:
More modifications of the print page as pdf keyword

emanlove committed:
Made corrections to the Print Page As PDF code and updated tests

Had some bad sloppy code within the keyword functionality. Cleaned that up
and added some tests.

emanlove committed:
Cleanup of commented out code

emanlove committed:
Merge branch 'master' into print-page-as-pdf-1824

emanlove committed:
Updated number of keywords

aaltat commented on atest/acceptance/keywords/print_page.robot:
```diff
@@ -0,0 +1,24 @@
+Test Setup    Remove Files    ${OUTPUTDIR}/selenium-page-*.pdf
+
+*** Test Cases ***
+Print Page As PDF Without Print Options
```
Is this same as next test?

emanlove replied:
I guess it is but not not intended to be. The first was intended to make sure that a page would print if no arguments were passed to it and iusing some of the default settings.

The second one is to check the automatic indexing of the saved filename. But you are right that a single test would check both of these scenarios.

emanlove replied:
Updated the second test using various arguments to make it sloightly different than the first.

emanlove committed:
Switched up print page scenarios so first test was not a repeat of the second
